src/ert/ensemble_evaluator/evaluator_async.py

```diff
@@ -0,0 +1,474 @@
+                await func(events)
+
+            self._batch_processing_queue.task_done()
+
+    async def _batch_events_into_buffer(self) -> None:
+        event_handler = {}
+
+        def set_event_handler(event_types: Set[str], function: Any) -> None:
+            for event_type in event_types:
+                event_handler[event_type] = function
+
+        for e_type, f in (
+            (EVGROUP_FM_ALL, self._fm_handler),
+            ({EVTYPE_ENSEMBLE_STARTED}, self._started_handler),
+            ({EVTYPE_ENSEMBLE_STOPPED}, self._stopped_handler),
+            ({EVTYPE_ENSEMBLE_CANCELLED}, self._cancelled_handler),
+            ({EVTYPE_ENSEMBLE_FAILED}, self._failed_handler),
+        ):
+            set_event_handler(e_type, f)
+
+        while True:
+            batch: List[
+                Tuple[Callable[[List[CloudEvent]], Awaitable[None]], CloudEvent]
+            ] = []
+            start_time = asyncio.get_running_loop().time()
+            while (
+                len(batch) < self._max_batch_size
+                and asyncio.get_running_loop().time() - start_time
+                < self._batching_interval
+            ):
+                try:
+                    event = await asyncio.wait_for(self._events.get(), timeout=0.1)
+                    function = event_handler[event["type"]]
+                    batch.append((function, event))
+                    self._events.task_done()
+                except asyncio.TimeoutError:
+                    continue
+            await self._batch_processing_queue.put(batch)
+
+    async def _fm_handler(self, events: List[CloudEvent]) -> None:
+        await self._append_message(self.ensemble.update_snapshot(events))
+
+    async def _started_handler(self, events: List[CloudEvent]) -> None:
+        if self.ensemble.status != ENSEMBLE_STATE_FAILED:
+            await self._append_message(self.ensemble.update_snapshot(events))
+
+    async def _stopped_handler(self, events: List[CloudEvent]) -> None:
+        if self.ensemble.status != ENSEMBLE_STATE_FAILED:
+            self._result = events[0].data  # normal termination
+            max_memory_usage = -1
+            for job in self.ensemble.snapshot.get_all_forward_models().values():
+                memory_usage = job.max_memory_usage or "-1"
+                if int(memory_usage) > max_memory_usage:
+                    max_memory_usage = int(memory_usage)
+            logger.info(
+                f"Ensemble ran with maximum memory usage for a single realization job: {max_memory_usage}"
+            )
+            await self._append_message(self.ensemble.update_snapshot(events))
+
+    async def _cancelled_handler(self, events: List[CloudEvent]) -> None:
+        if self.ensemble.status != ENSEMBLE_STATE_FAILED:
+            await self._append_message(self.ensemble.update_snapshot(events))
+            await self._stop()
+
+    async def _failed_handler(self, events: List[CloudEvent]) -> None:
```

Comment:
Probably for a different PR, but do you think events has to be of type list, or could/should it be a more general sequence?

Reply:
It could but that is indeed a different PR.